File: cosmos-master/code/sorting/circle_sort/circle_sort.c
/*Part of Cosmos by OpenGenus Foundation*/

#include <stdio.h>
#include <stdlib.h>
 
void swap(int *p, int *q)
{
    int tmp = *p;
    *p = *q;
    *q = tmp;
}

void print_vector(int *a, int n)
{
    for (int i = 0; i < n; i++)
        printf("%d ", a[i]);
    printf("\n");
}

int circle_sort(int *a, int n, int lower, int upper, int swaps)
{
    if (lower == upper)
        return swaps;

    int low = lower;
    int high = upper;
    int mid = (upper - lower) / 2;

    while (lower < upper) {
        if (a[lower] > a[upper]) {
            swap(&a[lower], &a[upper]);
            swaps++;
        }
        lower++;
        upper--;
    }

    if (lower == upper) {
        if(a[lower] > a[upper + 1]) {
            swap(&a[lower], &a[upper + 1]);
            swaps++;
        }
    }

    circle_sort(a, n, low, low + mid, swaps);
    circle_sort(a, n, low + mid + 1, high, swaps);

    return swaps;
}

int main()
{   
    int n;
    int *a;
    printf("Inform the size of the array:\n");
    scanf("%d", &n);

    a = (int *) malloc(n * sizeof(int));

    printf("Enter elements of the array one by one\n");
    for (int i = 0; i < n; i++) {
        scanf("%d", &a[i]);
    }    

    printf("Unsorted List: ");
    print_vector(a, n);
    circle_sort(a, n, 0, n - 1, 0);
    printf("Sorted List: ");
    print_vector(a, n);

    free(a);
    return 0;
}

File: cosmos-master/code/sorting/circle_sort/circle_sort.py
# Python program to implement circle sort

# function to swap 2 elements of a list
def swap(a, i, j):
    temp = a[i]
    a[i] = a[j]
    a[j] = temp

# recursive function to perfrom circle sort on the given list
def circleSort(a, lower, upper, swaps):
    # base case
    if(lower == upper):
        return swaps

    low = lower
    high = upper
    mid = int((upper-lower)/2)


    while lower<upper:
        if(a[lower] > a[upper]):
            swap(a, lower, upper)
            swaps += 1
            print(a, swaps)

        lower += 1
        upper -= 1

    if (lower == upper):
        if(a[lower] > a[upper+1]):
            swap(a, lower, upper+1)
            swaps += 1
            print(a, swaps)

    circleSort(a, low, low+mid, swaps)
    circleSort(a, low+mid+1, high, swaps)

    return swaps

# hard coded driver function to run the program
def main():
    a = [6,5,3,1,8,7,2,4]
    print("Unsorted List: ", a)
    no_of_swaps = circleSort(a, 0, len(a)-1, 0)
    print("Sorted List: ", a)


# execute isOrdered() function
if __name__ == '__main__':
    main()
